### ISSUE TITLE
The potential bug of the amber_file_parser.py when reading Nonbonded parameters of amber topology file

### ISSUE THREAD

**Issue by phloglucinol** [openmm/openmm#4717]
In the original `openmm\Lib\site-packages\openmm\app\internal\amber_file_parser.py`, there is a function called `getNonbondTerms`. Its content is as follows (see the image below):

![original_code](https://github.com/user-attachments/assets/6b260b38-b759-46a5-b111-14a54d1b2491)

```py
numTypes = self.getNumTypes()
atomTypeIndexes=self._getAtomTypeIndexes()
type_parameters = [(0, 0) for i in range(numTypes)]
```

In this code, `type_parameters` is defined as a list. Later, when accessing the elements of `type_parameters`, it is assumed that the indices of `type_parameters` are the values contained in `atomTypeIndexes`, and since `type_parameters` is a list, its indices must be consecutive integers.

For example, in the following double for loop:
```py
for i in range(numTypes):
    for j in range(numTypes):
        index = int(self._raw_data['NONBONDED_PARM_INDEX'][numTypes*i+j]) - 1
        if index < 0: continue
        rij = type_parameters[i][0] + type_parameters[j][0]
        wdij = sqrt(type_parameters[i][1] * type_parameters[j][1])
```

Using `range(numTypes)` also assumes that the unique elements in `atomTypeIndexes` are consecutive integers starting from 1.

This assumption works well for most Amber topology files. However, I recently encountered an Amber topology file modified by ParmEd, where the content of `%FLAG ATOM_TYPE_INDEX` is as follows (see the image below):

`ATOM_TYPE_INDEX` is actually the `atomTypeIndexes` parsed by `amber_file_parser.py`. Of course, the screenshot does not show all the content, but the subsequent `15 16 16` represent water molecules, and there are no other types of residues afterward. Therefore, the atom types shown in the screenshot include all the atom types in the current system, with a total of 15 atom types. However, we notice that there are no atoms of type 13 in this topology.

![special_amber_topology](https://github.com/user-attachments/assets/a7728791-cd1c-4242-a711-5921df0b9ffb)

Moreover, when we examine the content of `%FLAG NONBONDED_PARM_INDEX`, it contains 16*16=256 numbers.

![amber_topology_pointers_flag](https://github.com/user-attachments/assets/13808552-221f-4cd7-bbaa-4deb1c2164f4)

This corresponds to the original topology having 16 atom types (as for why `%FLAG ATOM_TYPE_INDEX` is missing atoms of type 13, we are not sure at the moment).

Furthermore, the `numTypes` given in the `%FLAG POINTERS` of the original topology file is also 16. This implies that the current `NONBONDED_PARM_INDEX` may be an index table that includes the nonbonded parameters of atoms of type 13 with other atoms. Based on this, we may need to make the following modifications to `amber_file_parser.py`:

![modified_code](https://github.com/user-attachments/assets/14b493e7-ac6f-467a-949b-ea764e53bd03)

We introduce a variable `atomTypesValues` to extract all the "in-use" atom types from the topology, change the data type of `type_parameters` to a dictionary, and modify the corresponding double for loop statements. I am not certain if this modification is sufficient to handle the parameter reading of such slightly unusual but indeed existing Amber topology files.

Here is the topology file and the coordinate file。
[top_crd.zip](https://github.com/user-attachments/files/17659426/top_crd.zip)

**Comment by phloglucinol** [comment]
When creating a system object using the code below based on the aforementioned topology and coordinate files, the original `amber_file_parser.py` sets the van der Waals parameters of all particles in the `nonbonded_force` of the system to 1 (sigma) and 0 (epsilon).

```py
try:
    import openmm
    from openmm import unit
    from openmm import app
except ImportError:  # Openmm < 7.6
    from simtk import openmm
    from simtk import unit
    from simtk.openmm import app
top = "lM2B.prmtop"
crd = "lM2B.prmcrd"
prmtop = app.AmberPrmtopFile(top)
inpcrd = app.AmberInpcrdFile(crd)
constraints=app.HBonds
rigid_water = True
nonbondedCutoff = 9.0 * unit.angstroms # default cutoff distance
nonbondedMethod = app.PME
hydrogenMass = None
system = prmtop.createSystem(constraints=constraints, nonbondedMethod=nonbondedMethod, rigidWater=rigid_water, nonbondedCutoff=nonbondedCutoff, hydrogenMass=hydrogenMass)

forces = {system.getForce(index).__class__.__name__: system.getForce(index) for index in range(system.getNumForces())}

f = forces['NonbondedForce']
for p_idx in range(f.getNumParticles()):
    paras = f.getParticleParameters(p_idx)
    with open('./debug2.csv', 'a') as ff:
        ff.write(f'{p_idx},{paras}\n')
```

The reason for this behavior is that the original `amber_file_parser.py` treats the unique elements in `atomTypeIndexes` as consecutive numbers. As a result, the program reaches the code block marked with the red arrow in the image below and raises an `NbfixPresent` exception.

![NbfixPresent_raise](https://github.com/user-attachments/assets/c47054e6-f8eb-4448-afac-b925df453a9e)

The code block that handles the `NbfixPresent` exception is as follows:

![except_handle_of_NbfixPresent](https://github.com/user-attachments/assets/80fff6be-c9f8-4cd5-b108-4b0311ed9799)

**Comment by peastman** [comment]
It isn't clear to me exactly what problem you're describing.  The values in `ATOM_TYPE_INDEX` should be integers between 1 and `NTYPES`.  And that seems to be the case.  `NTYPES` is 16.  All the values in `ATOM_TYPE_INDEX` are between 1 and 16.  It happens there are no atoms with type 13, but is that a problem?  Is there any code that assumes every atom type must be used by an atom?

**Comment by phloglucinol** [comment]
I apologize for not clearly explaining the potential issue. My point is that in the original `amber_file_parser.py`, there is a for loop (see the image below) that iterates based on `range(self.getNumAtoms())`.
![first_for_loop](https://github.com/user-attachments/assets/9ea92ac5-a137-401c-9335-4e6b5410a585)

Since `atomTypeIndexes` contains only 15 atom types, this loop will not parse the information corresponding to atoms of type 13. At this step, the LJ parameters of all atoms are parsed and stored in a list called `type_parameters`. For the topology file I am working with, `numTypes` is 16, but there are actually only 15 unique numbers representing atom types in the system's `atomTypeIndexes` (all consecutive numbers from 1 to 16, but missing 13). As a result, the 13th element (index 12) of `type_parameters` will not be updated and will retain its initial value of (0, 0).

Subsequently, in the double for loop code that checks for any off-diagonal modified LJ terms that would require an NBFIX-like solution:

![second_loop](https://github.com/user-attachments/assets/f51ad4fd-1920-4d2a-888b-00c56765a0a8)


For each possible pair of atoms, the LJ parameters of each atom need to be extracted to calculate `rij` and `wdij`. At this point, the for loop iterates based on `range(numTypes)`, which will inevitably attempt to read the LJ parameters of atoms of type 13. However, in the previous `type_parameters`, the LJ parameters of atoms of type 13 were not updated and remain as (0, 0). Despite this, `NONBONDED_PARM_INDEX` still contains 16*16=256 elements and is capable of accessing all nonbonded parameters related to atom type 13.

In other words, in the image below, `a` and `b` are not zero, and the `elif` branch will be taken.

![raise_NbfixPresent_](https://github.com/user-attachments/assets/ba4f0306-f619-46f5-a0d7-7efc13190e40)

However, at this point, the `a` and `b` values read from `NONBONDED_PARM_INDEX` will not correspond to the incorrectly assigned LJ parameters of atom type 13, and the `NbfixPresent` exception in the `elif` block will be raised.

**Comment by peastman** [comment]
Is that a problem?  The exception gets caught and tells it to treat the LJ parameters as NBFix terms:

https://github.com/openmm/openmm/blob/53770948682c40bd460b39830d4e0f0fd3a4b868/wrappers/python/openmm/app/internal/amber_file_parser.py#L945-L948

It still works correctly.  It just means it gets implemented in a different way.

**Comment by phloglucinol** [comment]
Thank you for your reply. I initially focused on this part of the code in `amber_file_parser.py` because I was performing free energy calculations for the annihilation of small molecule ligands based on alchemical transformations. I used the `create_alchemical_system` function from the `alchemy.AbsoluteAlchemicalFactory` class in the `alchemy.py` module of `openmmtools` to create the system. After reading and creating the system using the `test.prmtop` and `test.prmcrd` files I provided, I found that I could not control the LJ potential energy of the ligand molecule using λ.

Indeed, when the `NbfixPresent` exception is caught, `amber_file_parser.py` re-parses the system's nonbonded parameters and handles the calculation of the van der Waals potential energy for the LJ term using a `CustomNonbondedForce`:

```py
try:
    nonbondTerms = prmtop.getNonbondTerms()
except NbfixPresent:
    nbfix = True
    numTypes = prmtop.getNumTypes()
    parm_acoef = [float(x) for x in prmtop._raw_data['LENNARD_JONES_ACOEF']]
    parm_bcoef = [float(x) for x in prmtop._raw_data['LENNARD_JONES_BCOEF']]
    nbidx = [int(x) for x in prmtop._raw_data['NONBONDED_PARM_INDEX']]
    acoef = [0 for i in range(numTypes*numTypes)]
    bcoef = acoef[:] # copy
    ene_conv = units.kilocalories_per_mole.conversion_factor_to(units.kilojoules_per_mole)
    length_conv = units.angstroms.conversion_factor_to(units.nanometers)
    afac = sqrt(ene_conv) * length_conv**6
    bfac = ene_conv * length_conv**6
    for i in range(numTypes):
        for j in range(numTypes):
            idx = nbidx[numTypes*i+j] - 1
            if idx < 0: continue
            acoef[i+numTypes*j] = sqrt(parm_acoef[idx]) * afac
            bcoef[i+numTypes*j] = parm_bcoef[idx] * bfac
    if has_1264:
        cfac = ene_conv * length_conv**4
        ccoef = [0 for i in range(numTypes*numTypes)]
        for i in range(numTypes):
            for j in range(numTypes):
                idx = nbidx[numTypes*i+j] - 1
                if idx < 0: continue
                ccoef[i+numTypes*j] = parm_ccoef[idx] * cfac
        cforce = mm.CustomNonbondedForce('(a/r6)^2-b/r6-c/r^4; r6=r^6;'
                                         'a=acoef(type1, type2);'
                                         'b=bcoef(type1, type2);'
                                         'c=ccoef(type1, type2);')
    else:
        cforce = mm.CustomNonbondedForce('(a/r6)^2-b/r6; r6=r^6;'
                                         'a=acoef(type1, type2);'
                                         'b=bcoef(type1, type2);')
    cforce.addTabulatedFunction('acoef',
                mm.Discrete2DFunction(numTypes, numTypes, acoef))
    cforce.addTabulatedFunction('bcoef',
                mm.Discrete2DFunction(numTypes, numTypes, bcoef))
    if has_1264:
        cforce.addTabulatedFunction('ccoef',
                    mm.Discrete2DFunction(numTypes, numTypes, ccoef))
    cforce.addPerParticleParameter('type')
    for atom in prmtop._getAtomTypeIndexes():
        cforce.addParticle((atom-1,))
```

This works fine for regular MD. However, it is possible that the `_alchemically_modify_CustomNonbondedForce` function corresponding to `CustomNonbondedForce` has not been fully implemented when assigning the creation functions for the alchemical forces corresponding to different types of forces in the `create_alchemical_system` function of the `alchemy.AbsoluteAlchemicalFactory` class in `openmmtools` (as shown in the image below).

![_alchemically_modify_CustomNonbondedForce](https://github.com/user-attachments/assets/c2271199-9450-4ad8-ac70-82e4c1bca548)

This could be the real reason why I encountered the issue of the LJ potential energy not being controlled by λ_vdw in my alchemical simulations. We plan to change the topology by deleting the nonbonded parameters related to atom type 13 from the original topology to ensure the correspondence between `atomTypeIndexes`, `numTypes`, and `NONBONDED_PARM_INDEX`. This may also eliminate the need to modify the original OpenMM code. Thank you for your response again.
